src/fstext/determinize-star-inl.h

```diff
@@ -468,99 +469,221 @@ template<class Arc> class DeterminizerStar {
+    // e.g. if the original set is (s1, s2)
+    // s1 -eps-> s4
+    // s2 -eps-> s3 -eps-> s4
+    // then the depth for s4 would eventually be 2
+
+    // to avoid duplicate of computation, we use a priority_queue for storing
+    // the state_id's to process, and always pick the one with the smallest
+    // depth so far first
+    //
+    // this would be similar to topo-sort the states before processing;
+    // (not quite identical since we don't compute the definitive depth first,
+    // but gradually update the depth while running
+    // it seems to speed up the process very significantly already
+    int depth;
+  };
+
+  // this is the struct that is put into the priority_queue. 
+  // The comparison operator makes sure that the queue.top() returns
+  // the one with the least depth
+  struct IdAndDepth {
+    IdAndDepth (InputStateId i, int d):
+      id(i), depth(d) {}
+    InputStateId id;
+    int depth;
+    bool operator <(const IdAndDepth& other) const {
+      return this->depth > other.depth;  // we want the lowest depth on top
+    }
+  };
+  
+  // queue for EpsilonClosure() function
+  // put it here so the memeory allocation for better effiency
+  // otherwise it gets created and destroyed constantly, also
+  // possibly a lot of resize() call the for underlying vector
+  priority_queue<IdAndDepth> queue;
+
+  // Sub-routine that we call in EpsilonClosure()
+  // To avoid the function getting too long, we make it a separate function
+  // It takes the current "unprocessed_weight" and propagate it to the 
+  // states accessible from elem.state by an epsilon arc
+  // and add the results to cur_subset.
+  void ExpandOneElement(const Element &elem,
+                        bool sorted,
+                        Weight unprocessed_weight,
+                        std::map<InputStateId, EpsilonClosureInfo> *cur_subset,
+                        int this_depth) {
+    // now we are going to propagate the "unprocessed_weight"
+    for (ArcIterator<Fst<Arc> > aiter(*ifst_, elem.state);
+         !aiter.Done(); aiter.Next()) {
+      const Arc &arc = aiter.Value();
+      if (sorted && arc.ilabel > 0) {
+        break;
+        // Break from the loop: due to sorting there will be no
+        // more transitions with epsilons as input labels.
+      }
+      if (arc.ilabel == 0) {  // Epsilon transition.
+        Element next_elem;
+        next_elem.state = arc.nextstate;
+        next_elem.weight = Weight::Zero();
+        Weight next_unprocessed_weight
+                       = Times(unprocessed_weight, arc.weight);
+
+        // now must append strings
+        if (arc.olabel == 0)
+          next_elem.string = elem.string;
+        else {
+          vector<Label> seq;
+          repository_.SeqOfId(elem.string, &seq);
+          if (arc.olabel != 0)
+            seq.push_back(arc.olabel);
+          next_elem.string = repository_.IdOfSeq(seq);
+        }
```

Comment:
Oh - disappointing.
Make sure you address that iterator invalidation problem I pointed out, if
you haven't already.

I realized after I commented that there would be a penalty in that method
because it first does a find() and then uses operator [] to set the
element, and both are n log n.
I have a couple of ideas for how you could address this:
(A) [written before idea (B)]
Anyway, having a separate function like this is probably a better way to
structure it, and we can do further optimizations, and then go back to the
old way of doing it if it doesn't work (but maybe keep the separate
function, inlined, for use just once).
My idea is to optimize that [newly created] method is this:
Instead of 'find' you can use function lower_bound to find the first
element which is <= the given element.  Then if the contents are the thing
you want, you found it; otherwise you can use operator -- on the returned
iterator (assuming it doesn't return the map's begin()), and use the result
as a 'hint' in the version of 'insert' that takes the hint.  This will
avoid having two n log n operations.

Idea (B) [I think this is better]:
Instead of a map from InputStateId to EpsilonClosureInfo, you can have a
vector of EpsilonClosureInfo, and separately store a map from InputStateId
to the index into the EpsilonClosureInfo vector.  The map from InputStateId
to the index can be a std::vector<int32> that is allocated just once, in
the constructor.  You will never 'clean up' this vector: the approach will
be that you do a direct lookup, then test 'is this index valid' (i.e. is it
in-range and if so does it point to what we asked for).  If not, you treat
it as empty.
The queue could then be a queue of pairs (depths, indexes) (i.e. store the
indexes into the EpsilonClosureInfo vector), which means no map lookup is
required.
The vector of EpsilonClosureInfo could also be a class member variable, to
avoid allocation and deallocation.


Dan


On Sun, Oct 11, 2015 at 3:41 PM, hainan-xv <notifications@github.com> wrote:

> In src/fstext/determinize-star-inl.h
> <https://github.com/kaldi-asr/kaldi/pull/208#discussion_r41714547>:
>
> > +        Element next_elem;
> > +        next_elem.state = arc.nextstate;
> > +        next_elem.weight = Weight::Zero();
> > +        Weight next_unprocessed_weight
> > +                       = Times(unprocessed_weight, arc.weight);
> > +
> > +        // now must append strings
> > +        if (arc.olabel == 0)
> > +          next_elem.string = elem.string;
> > +        else {
> > +          vector<Label> seq;
> > +          repository_.SeqOfId(elem.string, &seq);
> > +          if (arc.olabel != 0)
> > +            seq.push_back(arc.olabel);
> > +          next_elem.string = repository_.IdOfSeq(seq);
> > +        }
>
> Just implemented this - it seems it's way slower than before: around 40
> seconds, while the initial implementation is about 20s slower. Also now I
> am using const & in the constructor for EpsilonClosureInfo() which should
> bring some speed up...
>
> It seems to me this algorithm is doing the exact same thing as our 1st
> implementation; except that in the first implementation we could utilize
> the fact that the StateId's are sorted so the insertion to the set is
> linear, while here it is n log n. I just pushed the changes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/208/files#r41714547>.
>